hail/src/main/scala/is/hail/nativecode/NativeCodeNode.scala

```diff
@@ -0,0 +1,201 @@
+package is.hail.nativecode
+
+import is.hail.utils.ArrayBuilder
+
+trait NativeCodeUnit {
+  def toString: String
+}
+
+abstract class NativeCodeStatement extends NativeCodeUnit {
+
+  def toString: String
+
+}
+
+class NativeCodeVariable(name: String, t: String, init: NativeCodeStatement = null) extends Serializable {
+
+  def declare: String = s"$t $name${if (init == null) "" else s" = $init"};\n"
+
+  def load(): NativeCodeStatement = NCode(name)
+
+  def store(v: NativeCodeStatement): NativeCodeStatement = NCode(s"$name = $v")
+
+  override def toString: String = name
+}
+
+object NCode {
+  def apply(expr: String): NativeCodeStatement = new NativeCodeStatement {
+      override def toString: String = expr
+  }
+}
+
+trait NativeScope extends NativeCodeUnit {
+
+  private val refs: ArrayBuilder[NativeCodeVariable] = new ArrayBuilder[NativeCodeVariable]()
```

Comment:
So C++ lets you define variables at any point in a block.  I think that should be the default.  If we need top of block definitions (I'm not sure we do), we should have a separate function for that.